User-facing documentation for Logstash on ECK (#6743)
elastic/cloud-on-k8s: docs/orchestrating-elastic-stack-applications/logstash.asciidoc

            password => "${QS_ES_PASSWORD}"
            cacert => "${QS_ES_SSL_CERTIFICATE_AUTHORITY}"
          }
        }
----

Alternatively, it can be provided through a Secret specified in the `spec.pipelinesRef` element. The Secret must have a `logstash.yml` entry with this configuration:
[source,yaml,subs="attributes,+macros"]
----
apiVersion: logstash.k8s.elastic.co/v1alpha1
kind: Logstash
metadata:
  name: quickstart
spec:
  version: {version}
  count: 1
  elasticsearchRefs:
    - clusterName: qs
      name: quickstart
  pipelinesRef:
    secretName: quickstart-pipeline
---
apiVersion: v1
kind: Secret
metadata:
  name: quickstart-pipeline
stringData:
  pipelines.yml: |-
    - pipeline.id: main
      config.string: |
        input {
          beats {
            port => 5044
          }
        }
        output {
          elasticsearch {
            hosts => [ "${QS_ES_HOSTS}" ]
            user => "${QS_ES_USER}"
            password => "${QS_ES_PASSWORD}"
            cacert => "${QS_ES_SSL_CERTIFICATE_AUTHORITY}"
          }
        }

----

Logstash on ECK will support all options present in `pipelines.yml`, including settings to update the number of workers, and

>  the size of the batch that the pipeline will process. This also includes using `path.config` to point to volumes

Author: I want to point out `path.config` specifically - it provides another way to store pipeline configurations, rather than inline in the `pipeline`, or the secret pointed to in the `pipelineRef` - it's a way to have more complex pipelines than is possible via `pipeline` or `pipelineRef` due to limitations on k8s `Secret` and `ConfigMap` definitions

 mounted on the Logstash container:

[source,yaml,subs="attributes,+macros,callouts"]
----
apiVersion: logstash.k8s.elastic.co/v1alpha1
kind: Logstash
metadata:
  name: quickstart
spec:
  version: {version}
  count: 1
  elasticsearchRefs:
    - clusterName: qs
      name: quickstart
  pipelines:
    - pipeline.id: main
      config.string: |
        input {
          beats {
            port => 5044
          }
        }
        output {
          elasticsearch {
            hosts => [ "${QS_ES_HOSTS}" ]
            user => "${QS_ES_USER}"
            password => "${QS_ES_PASSWORD}"
            cacert => "${QS_ES_SSL_CERTIFICATE_AUTHORITY}"
          }
        }
----

NOTE: Persistent Queues and Dead Letter Queues are not currently supported.

[id="{p}-logstash-pipelines-es"]
=== Using Elasticsearch in Logstash Pipelines

> The `spec.elasticsearchRefs` section provides a mechanism to help configure Logstash to estabish a secured connection to one or more managed Elasticsearch clusters. By default, it targets all nodes in your cluster. If you want to direct traffic to specific nodes of your Elasticsearch cluster, refer to <<{p}-traffic-splitting>> for more information and examples.

Author: What do you think of 
```suggestion
The `spec.elasticsearchRefs` section provides a mechanism to help configure Logstash to estabish a secured connection to one or more managed Elasticsearch clusters. By default, each `elasticsearchRef` will target all nodes in its referenced Elasticsearch cluster. If you want to direct traffic to specific nodes of your Elasticsearch cluster, refer to <<{p}-traffic-splitting>> for more information and examples.
```

> The `spec.elasticsearchRefs` section provides a mechanism to help configure Logstash to estabish a secured connection to one or more managed Elasticsearch clusters. By default, it targets all nodes in your cluster. If you want to direct traffic to specific nodes of your Elasticsearch cluster, refer to <<{p}-traffic-splitting>> for more information and examples.

Reviewer: Much more precise and therefore informative.  I like it with one small tweak. Let's keep it in present tense. 

By default, each `elasticsearchRef` ~will~ target**+s** all nodes in its referenced Elasticsearch

> To use `elasticsearchRefs` in a Logstash pipeline, the Logstash operator will create the necessary resources from the associated Elasticsearch, and provides environment variables to allow these resources to be accessed from the pipeline configuration, and will be replaced at runtime with the appropriate values.

Author: I think I prefer cluster here - I think it less ambiguous than instance, which I could potentially mean a node or a cluster
```suggestion
When you use `elasticsearchRefs` in a Logstash pipeline, the Logstash operator creates the necessary resources from the associated Elasticsearch cluster, and provides environment variables to allow these resources to be accessed from the pipeline configuration. 
Environment variables are replaced at runtime with the appropriate values.```

> The environment variables have a fixed naming convention:

Author: Done

            password => "${PROD_ES_ES_PASSWORD}"
            cacert => "${PROD_ES_ES_SSL_CERTIFICATE_AUTHORITY}"
          }
          elasticsearch {   <4>
            hosts => [ "${QA_ES_ES_HOSTS}" ]
            user => "${QA_ES_ES_USER}"
            password => "${QA_ES_ES_PASSWORD}"
            cacert => "${QA_ES_ES_SSL_CERTIFICATE_AUTHORITY}"
          }
        }

----

<1> Define Elasticsearch references in the CRD. This will create the appropriate Secrets to store certificate details and the rest of the connection information, and create environment variables to allow them to be referred to in Logstash pipeline configurations.
<2> This refers to an Elasticsearch cluster residing in the same namespace as the Logstash instances.
<3> This refers to an Elasticsearch cluster residing in a different namespace to the Logstash instances.
<4> Elasticsearch output definitions - use the environment variables created by the Logstash operator when specifying an `ElasticsearchRef`. Note the use of "normalized" versions of the `clusterName` in the environment variables used to populate the relevant fields.

[id="{p}-logstash-expose-services"]
=== Expose Services

> By default, the Logstash operator creates a headless Service for the metrics endpoint to enable metric collection by the Metricbeat sidecar for Stack Monitoring:

Author: `Service` is a specific k8s resource, and I think it should be capitalized to show that

[source,sh]
----
kubectl get service quickstart-ls-api
----

[source,sh,subs="attributes"]
----
NAME                TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)    AGE
quickstart-ls-api   ClusterIP   None         <none>        9600/TCP   48s
----

> Additional services can be added by using the `spec.services` section of the resource:

Author: What do you think of using "in" here? The `spec.services` refers to a configuration stanza in the resource description, and we add services by making changes here

```suggestion
Additional services can be added in the `spec.services` section of the resource:
```

        - name: LS_JAVA_OPTS
          value: "-Xmx2g -Xms2g"
        resources:
          requests:
            memory: 1Gi
            cpu: 0.5
          limits:
            memory: 4Gi
            cpu: 2
----

The name of the container in the Pod template must be `logstash`.

[id="{p}-logstash-configuration-examples"]
== Configuration examples

This section contains manifests that illustrate common use cases, and can be your starting point in exploring Logstash deployed with ECK. These manifests are self-contained and work out-of-the-box on any non-secured Kubernetes cluster. They all contain a three-node Elasticsearch cluster and a single Kibana instance.

CAUTION: The examples in this section are for illustration purposes only and should not be considered to be production-ready. Some of these examples use the `node.store.allow_mmap: false` setting on Elasticsearch which has performance implications and should be tuned for production workloads, as described in <<{p}-virtual-memory>>.

=== Single Pipeline defined in CRD

[source,sh,subs="attributes"]
----
kubectl apply -f {logstash_recipes}/logstash-eck.yaml
----

Deploys Logstash with a single pipeline defined in the CRD

> === Single Pipeline defined in Secret

Author: ```suggestion
=== Single pipeline defined in Secret
```

Back to the "reserved" k8s resources again...

> === Single Pipeline defined in Secret

Reviewer: Maybe call it out as `Pipeline` so it's obvious that we're not speaking of a generic Logstash pipeline? 
You make the call either way.

> === Single Pipeline defined in Secret

Author: Let's keep it as is

          }
        }
        output {
          elasticsearch {
            hosts => [ "${QS_ES_HOSTS}" ]
            user => "${QS_ES_USER}"
            password => "${QS_ES_PASSWORD}"
            cacert => "${QS_ES_SSL_CERTIFICATE_AUTHORITY}"
          }
        }

----

Logstash on ECK supports all options present in `pipelines.yml`, including settings to update the number of workers, and
 the size of the batch that the pipeline will process. This also includes using `path.config` to point to volumes
 mounted on the Logstash container:

[source,yaml,subs="attributes,+macros,callouts"]
----
apiVersion: logstash.k8s.elastic.co/v1alpha1
kind: Logstash
metadata:
  name: quickstart
spec:
  version: {version}
  count: 1
  elasticsearchRefs:
    - clusterName: qs
      name: quickstart
  pipelines:
    - pipeline.id: main
      config.string: |
        input {
          beats {
            port => 5044
          }
        }
        output {
          elasticsearch {
            hosts => [ "${QS_ES_HOSTS}" ]
            user => "${QS_ES_USER}"
            password => "${QS_ES_PASSWORD}"
            cacert => "${QS_ES_SSL_CERTIFICATE_AUTHORITY}"
          }
        }
----

> NOTE: Logstash persistent queues (PQs) and dead letter queues (DLQs) are not currently supported.

Author: Yes, I think that makes sense, I'll reword it

[id="{p}-logstash-configuration-pipeline-volume"]
=== Pipeline configuration in mounted volume

[source,sh,subs="attributes"]
----
kubectl apply -f {logstash_recipes}/logstash-pipeline-as-volume.yaml
----

Deploys Logstash with a single pipeline defined in a secret, mounted as a volume, and referenced by
`path.config`

[id="{p}-logstash-configuration-custom-index"]
=== Writing to a custom Elasticsearch index

[source,sh,subs="attributes"]
----
kubectl apply -f {logstash_recipes}/logstash-es-role.yaml
----

Deploys Logstash and Elasticsearch, and creates an updated version of the `eck_logstash_user_role` to write to a user specified index.

[id="{p}-logstash-configuration-stack-monitoring"]
=== Elasticsearch and Kibana Stack Monitoring

[source,sh,subs="attributes"]
----
kubectl apply -f {logstash_recipes}/logstash-monitored.yaml
----

Deploys an Elasticsearch and Kibana monitoring cluster, and a Logstash that will send its monitoring information to this cluster. You can view the stack monitoring information in the monitoring cluster's Kibana

[id="{p}-logstash-configuration-multiple-pipelines"]

> === Multiple pipelines/multiple es clusters

Reviewer 2: ```suggestion
=== Multiple pipelines/multiple Elasticsearch clusters
```

[source,sh,subs="attributes"]
----
kubectl apply -f {logstash_recipes}/logstash-multi.yaml
----

> Deploys Elasticsearch in prod and qa configurations, running in separate namespaces. Logstash is configured with a multiple pipeline->pipeline configuration, with a source pipeline routing to prod and qa pipelines.

Reviewer 2: ```suggestion
Deploys Elasticsearch in prod and qa configurations, running in separate namespaces. Logstash is configured with a multiple pipeline->pipeline configuration, with a source pipeline routing to `prod` and `qa` pipelines.
```
